[WIP] Create beginner’s tutorial for using Blitz (#8)
blitz-js/drafts: TUTORIAL.md

From the command line, `cd` into the directory where you’d like to store your code, and then run the following command:

```sh
blitz new mysite
```

This should create a `mysite` directory in your current directory.

Let’s look at what `blitz new` created:

```
mysite/
  app/
    components/
      ErrorBoundary.tsx
    layouts/
    pages/
      _app.tsx
      _document.tsx
      index.tsx
  db/
  integrations/
  jobs/
  node_modules/
  public/
  utils/
  .babelrc.js
  .env
  .eslintrc.js
  .gitignore
  .npmrc
  .prettierignore
  blitz.config.js
  package.json
  README.md
  tsconfig.json
```

These files are:

- The `app/` directory is a container for most of your project. This is where you’ll put any pages or api routes.

- `db`/ is where your database configuration goes. If you’re writing models or checking migrations, this is where to go.

- `node_modules/` is where your “dependencies” are stored. This is an auto-generated directory, so you don’t have to worry too much about it.

- `public/` is a directory where you will put any static assets. If you have images, files, or videos which you want to use in your app, this is where to put them. 

> - `utils/ is a good place to put any share util files which you might use across different sections of your app.

Reviewer: ```suggestion
- `utils/` is a good place to put any share util files which you might use across different sections of your app.
```

[ info ]  waiting on http://localhost:3000 ...
[ info ]  bundled successfully, waiting for typecheck results...
[ wait ]  compiling ...
[ info ]  bundled successfully, waiting for typecheck results...
[ ready ] compiled successfully - ready on http://localhost:3000
```

Now that the server’s running, visit http://localhost:3000/ with your Web browser. You’ll see a welcome page, with the Blitz logo. It worked!

## Write your first page

Now that your development environment—a “project”—is set up, you’re ready to start building out the app. First, we’ll create your first page.

Open the file `app/pages/index.tsx` and put the following code in it:

```tsx
export default () => (
  <div>
    <h1>Hello, world!</h1>
  </div>
)
```

This is the simplest page possible in Blitz. To look at it, go back to your browser and go to http://localhost:3000. You should see your text appear! Try editing the `index.tsx` file, and make it your own! When you’re ready, move on to the next section.

## Database setup
Now, we’ll setup the database and create your first model.

Open up `db/schema.prisma`. It’s a configuration file which our default database engine Prisma uses.

> By default, the apps is created with SQLite. If you’re new to databases, or you’re just interested in trying Blitz, this is the easiest choice. Not that when starting your first project, you may want to use a more scalable database like PostgreSQL, to avoid the pains of switching your database down the road.

Reviewer: Typo: "Not that when starting"

## Creating models

Now we’ll define your models—essentially, your database layout, with additional metadata.

In `schema.prisma`, we’ll create two models: `Question`, and `Choice`. A `Question` has a question and a publication date. A `Choice` has two fields: the text of the choice and a vote count. Each has an id, and each `Choice` is associated with a `Question`.

Edit the `schema.prisma` file so it looks like this:

```
// (datasource and generator)

...

model Question {
  id          Int      @default(autoincrement()) @id
  text        String
  publishedAt DateTime
  choices     Choice[]
}

model Choice {
  id         Int      @default(autoincrement()) @id
  text       String
  votes      Int      @default(0)
  question   Question @relation(fields: [questionId], references: [id])
  questionId Int
}
```

> Now, we need to migrate our database. This is a way of telling it that you have edited your schema in some way. Run this, and enter something like `init db` when prompted for a migration name:

Reviewer: ```suggestion
Now, we need to migrate our database. This is a way of telling it that you have edited your schema in some way. Run the below command. When it asks you to enter a migration name you can enter anything you want, perhaps `
"init db":
```

```sh
$ blitz db migrate
```

## Playing with the API

> Now, let’s hop into the interactive Blitz shell and play around with the free API Blitz gives you. To invoke the Blitz shell, use this command:

Reviewer: ```suggestion
Now, let’s hop into the interactive Blitz shell and play around with the free API Blitz gives you. To invoke the Blitz console, use this command:
```

For consistency and less words to know

```sh
$ blitz console
```

Once you’re in the console, explore the Database API:

```sh
# No questions are in the system yet.
⚡ > await db.question.findMany()
[]

# Create a new Question.
⚡ > q = await db.question.create({data: {text: 'What’s new?', publishedAt: new Date()}})
{ id: 1, text: 'What’s new?', publishedAt: 2020-04-24T22:08:17.307Z }

# Now it has an ID.
⚡ > q.id
1

# Access model field values via JavaScript attributes.
⚡ > q.text
"What’s new?"

⚡ > q.publishedAt
2020-04-24T22:08:17.307Z

# Change values by using the update function
⚡ > q = await db.question.update({where: {id: 1}, data: {text: 'What’s up?'}})
{ id: 1, text: 'What’s up?', publishedAt: 2020-04-24T22:08:17.307Z }

# db.questions.findMany() displays all the questions in the database.
⚡ > await db.questions.findMany()
[
  { id: 1, text: 'What’s up?', publishedAt: 2020-04-24T22:08:17.307Z }
]

> ```

Reviewer: Top level await doesn't work right now :(

Author: Oh damn:/ It worked for me when I was writing this but maybe that’s a cross-platform thing? Just leaving this here for now since this is the intended behaviour of the console

Reviewer: Let's change this to use `.then` so we can go ahead and publish this asap.

      data: { votes: votes + 1 },
    })
    alert("Success!" + JSON.stringify(updated))
  } catch (error) {
    alert("Error creating question " + JSON.stringify(error, null, 2))
  }
}

return (

...
```

Finally, we’ll just tell our `button` to call that function!

```jsx
<button onClick={() => handleVote(choice.id, choice.votes)}>Vote</button>
```

Just to be sure, this is what all that should look like:

```jsx
import { Suspense } from "react"
import { Head, Link, useRouter, useQuery } from "blitz"
import getQuestion from "app/questions/queries/getQuestion"
import deleteQuestion from "app/questions/mutations/deleteQuestion"
import updateChoice from "app/questions/mutations/updateChoice"

export const Question = () => {
  const router = useRouter()
  const id = parseInt(router?.query.id as string)
  const [question] = useQuery(getQuestion, { where: { id } })

  const handleVote = async (id, votes) => {
    try {
      const updated = await updateChoice({
        where: { id },
        data: { votes: votes + 1 },
      })
      alert("Success!" + JSON.stringify(updated))
    } catch (error) {
      alert("Error creating question " + JSON.stringify(error, null, 2))
    }

>   }

Reviewer: Should this be moved out and above the Question function?

Author: I usually colocate handler functions within components but more than happy to change this if it’s just a question of code style(?)

Reviewer: Ok this is fine
